collections/src/main/scala/strawman/collection/immutable/ChampHashMap.scala

```diff
@@ -0,0 +1,601 @@
+package strawman
+package collection.immutable
+
+import collection.{IterableFactory, Iterator, MapFactory, StrictOptimizedIterableOps}
+import collection.Hashing.{computeHash, keepBits}
+
+import scala.annotation.unchecked.{uncheckedVariance => uV}
+import scala.{Any, AnyRef, Array, Boolean, Double, IndexOutOfBoundsException, Int, NoSuchElementException, None, Nothing, Option, SerialVersionUID, Serializable, Some, Tuple2, Unit, `inline`, math, sys, UnsupportedOperationException}
+import scala.Predef.{assert, require, ???}
+
+import java.lang.Integer.bitCount
+import java.lang.System.arraycopy
+
+import strawman.collection.mutable.{Builder, ImmutableBuilder}
+
+/** This class implements immutable maps using a Compressed Hash-Array Mapped Prefix-tree.
+  * See paper https://michael.steindorfer.name/publications/oopsla15.pdf for more details.
+  *
+  *  @tparam K      the type of the keys contained in this hash set.
+  *  @tparam V      the type of the values associated with the keys in this hash map.
+  *
+  *  @author  Michael J. Steindorfer
+  *  @version 2.13
+  *  @since   2.13
+  *  @define Coll `immutable.ChampHashMap`
+  *  @define coll immutable champ hash map
+  */
+
+@SerialVersionUID(2L)
+final class ChampHashMap[K, +V] private[immutable] (val rootNode: MapNode[K, V], val cachedJavaHashCode: Int, val cachedSize: Int)
+  extends Map[K, V]
+    with MapOps[K, V, ChampHashMap, ChampHashMap[K, V]]
+    with StrictOptimizedIterableOps[(K, V), Iterable /* ChampHashMap */, ChampHashMap[K, V]]
+    with Serializable {
+
+  def iterableFactory: IterableFactory[Iterable] = Iterable
+
+  def mapFactory: MapFactory[ChampHashMap] = ChampHashMap
+
+  protected[this] def mapFromIterable[K2, V2](it: collection.Iterable[(K2, V2)]): ChampHashMap[K2, V2] = ChampHashMap.from(it)
+
+  protected[this] def fromSpecificIterable(coll: collection.Iterable[(K, V)]): ChampHashMap[K, V] = ChampHashMap.from(coll)
+
+  protected[this] def newSpecificBuilder(): Builder[(K, V), ChampHashMap[K, V]] = ChampHashMap.newBuilder()
+
+  override def knownSize: Int = cachedSize
```

Comment:
You might also want to override `size` (the default implementation tests whether `knownSize` is defined or not, so you would just save that comparison…)

Reply:
Also, `isEmpty` and `nonEmpty` could be overridden for performance too (the default implementation creates an `Iterator` and then checks whether that iterator is empty or not)